Question - 1 
Implement 3 different CNN architectures with a comparison table for the MNSIT
dataset using the Tensorflow library.
Note -
1. The model parameters for each architecture should not be more than 8000
parameters
2. Code comments should be given for proper code understanding.
3. The minimum accuracy for each accuracy should be at least 96%


In [1]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

In [2]:
# Load MNIST dataset
(x_train, y_train), (x_test, y_test) = keras.datasets.mnist.load_data()

# Preprocess data
x_train = x_train.astype("float32") / 255
x_test = x_test.astype("float32") / 255


In [3]:
# Define model architectures

#first architecture
def create_model_1():
    model = keras.Sequential([
        layers.Input(shape=(28, 28)),
        layers.Reshape(target_shape=(28, 28, 1)),
        layers.Conv2D(filters=8, kernel_size=5, activation='relu'),
        layers.MaxPooling2D(pool_size=(2, 2)),
        layers.Conv2D(filters=16, kernel_size=5, activation='relu'),
        layers.MaxPooling2D(pool_size=(2, 2)),
        layers.Flatten(),
        layers.Dense(units=10, activation='softmax')
    ])
    return model


# Create models
model_1 = create_model_1()

# Compile models
model_1.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

# Train models
model_1.fit(x_train, y_train)

# Evaluate models
_, acc_1 = model_1.evaluate(x_test, y_test)


# Comparison table
print(f"{'Model':<10}{'Accuracy':<10}")
print(f"{'Model - 1':<10}{acc_1:<10.4f}")


# Check the number of model parameters for each architecture
params_1 = model_1.count_params()

# Print the number of model parameters for each architecture
print(f"{'Model':<10}{'Parameters':<10}")
print(f"{'Model 1':<10}{params_1:<10}")

313/313 [==============================] - 2s 6ms/step - loss: 0.0704 - accuracy: 0.9786
Model     Accuracy  
Model - 1 0.9786    
Model     Parameters
Model 1   5994      


In [4]:
# Second architecture
def create_model_2():
    model = keras.Sequential([
        layers.Input(shape=(28, 28)),
        layers.Reshape(target_shape=(28, 28, 1)),
        layers.Conv2D(filters=8, kernel_size=3, activation='relu'),
        layers.MaxPooling2D(pool_size=(2, 2)),
        layers.Conv2D(filters=16, kernel_size=3, activation='relu'),
        layers.MaxPooling2D(pool_size=(2, 2)),
        layers.Flatten(),
        layers.Dense(units=10, activation='softmax')
    ])
    return model


# Create models
model_2 = create_model_2()

# Compile models
model_2.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

# Train models
model_2.fit(x_train, y_train)

# Evaluate models
_, acc_2 = model_2.evaluate(x_test, y_test)

# Comparison table
print(f"{'Model':<10}{'Accuracy':<10}")
print(f"{'Model - 2':<10}{acc_2:<10.4f}")

# Check the number of model parameters for each architecture
params_2 = model_2.count_params()

# Print the number of model parameters for each architecture
print(f"{'Model':<10}{'Parameters':<10}")
print(f"{'Model 2':<10}{params_2:<10}")

313/313 [==============================] - 2s 5ms/step - loss: 0.1128 - accuracy: 0.9668
Model     Accuracy  
Model - 2 0.9668    
Model     Parameters
Model 2   5258      


In [5]:
# Third architecture
def create_model_3():
    model = keras.Sequential([
        layers.Input(shape=(28, 28)),
        layers.Reshape(target_shape=(28, 28, 1)),
        layers.Conv2D(filters=8, kernel_size=3, activation='relu'),
        layers.MaxPooling2D(pool_size=(2, 2)),
        layers.Conv2D(filters=16, kernel_size=3, activation='relu'),
        layers.MaxPooling2D(pool_size=(2, 2)),
        layers.Flatten(),
        layers.Dense(units=10, activation='softmax')
    ])
    return model


# Create models
model_3 = create_model_3()

# Compile models
model_3.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

# Train models
model_3.fit(x_train, y_train)

# Evaluate models
_, acc_3 = model_3.evaluate(x_test, y_test)

# Comparison table
print(f"{'Model':<10}{'Accuracy':<10}")
print(f"{'Model - 3':<10}{acc_3:<10.4f}")

# Check the number of model parameters for each architecture
params_3 = model_3.count_params()

# Print the number of model parameters for each architecture
print(f"{'Model':<10}{'Parameters':<10}")
print(f"{'Model 3':<10}{params_3:<10}")

313/313 [==============================] - 2s 6ms/step - loss: 0.0902 - accuracy: 0.9739
Model     Accuracy  
Model - 3 0.9739    
Model     Parameters
Model 3   5258      


## So, we can see that all three models have achieved an accuracy of at least 96% on the test data and model parameters do not exceed more than 8000. 
Model 1 has an accuracy of 97.86%, 

Model 2 has an accuracy of 96.68% and 

Model 3 has an accuracy of 97.39%.